### DISCUSSION TITLE
Trust Hashicorp Raft peers and don't overwrite with -master.peers

### DISCUSSION THREAD

**Discussion by jameshartig** [seaweedfs/seaweedfs#4380] [General]
It looks like on startup the hashicorp raft code call `s.UpdatePeers()` which overwrites the existing raft configuration with the peers command-line configuration. This is unfortunate because if a master is restarted before it's command-line arguments are updated then it'll remove a newly added master. Can we instead add a command-line switch around that? Maybe something like `-master.raft-force-peers`.

Is there a reason why we don't just trust the previous raft configuration that was saved on disk except when we're bootstrapping? I would think the only use-case for overwriting the raft state would be outage recovery and this is how Consul handles it: https://developer.hashicorp.com/consul/tutorials/datacenter-operations/recovery-outage#manual-recovery-using-peers-json. I could see the addition of a new flag (or making `-master.peers` optional if raft is bootstrapped) being similar to the peers.json idea from Consul. If there isn't an outage then I would assume you always want to trust the latest raft state.

We would like to bootstrap seaweed masters with a hardcoded list but then scale up/down the cluster without having to modify environment variables or command-line arguments.

/cc @kmlebedev

**Comment by jameshartig** [comment]
@kmlebedev Any thoughts here?

**Comment by kmlebedev** [comment]
@jameshartig I don't understand what actual problem are you solving?
Can you describe it step by step?

master.peers is needed for initialization

Further, the list can be managed manually through the `weed shell`
```
> cluster.raft.add --help
Usage of cluster.raft.add:
  -address string
    	server grpc address
  -id string
    	server id
  -voter
    	assign it a vote (default true)
> cluster.raft.remove --help
Usage of cluster.raft.remove:
  -id string
    	server id
> cluster.raft.ps --help
Usage of cluster.raft.ps:
```

Also, if suddenly one of the peers is unavailable via grpc for a long time, it will automatically be removed from the raft list.
So through -master.peers you can safely add new raft peers and not worry about deleting old ones

**Reply by jameshartig** [reply]
Hey @kmlebedev this is my main concern:

> We would like to bootstrap seaweed masters with a hardcoded list but then scale up/down the cluster without having to modify environment variables or command-line arguments.

We plan to use preemptible nodes for Seaweed and will frequently have nodes being removed and added. The problem is that we don't want to programmatic have to update the `-master.peers` command-line argument with the new set of servers every time a new server joins or leaves. If we do not update the argument then if a seaweed instance is restarted it'll revert the list of peers to the ones listed in the command-line argument.

I noted that this behavior is contrary to other raft-based services which only require the peer list to initialize the raft cluster the first time and then subsequent times it's ignored if there is already raft files on disk. The raft state stored on disk should be trusted over the command-line arguments unless you're specifically bootstrapping the cluster for the first time.

**Comment by kmlebedev** [comment]
There is also a complete overwrite mode.
```
weed master --help | grep raftBootstrap
Example: weed master -port=9333
Default Usage:
  -raftBootstrap
    	Whether to bootstrap the Raft cluster
```
Since s.UpdatePeers() only adds new peers

**Reply by jameshartig** [reply]
That's not my understanding of the code. Looking at https://github.com/seaweedfs/seaweedfs/blob/master/weed/server/raft_hashicorp.go#L81 it will check all the servers in the raft configuration and any that are not in `s.peers` (which comes from the peers command-line argument) will be removed. Including removing itself if itself isn't in `s.peers`.

**Comment by kmlebedev** [comment]
@jameshartig Yes, there are peer removals, but except when we explicitly update the cluster sequentially within 15 minutes after launch, control is transferred to the new leader

**Reply by jameshartig** [reply]
Yes, so if there is a new leader within the first 15 minutes of launch the cluster will be broken if the command-line argument does not match the current state of the cluster. That means if you do a rolling restart of the whole cluster you're guaranteed to have some leadership transfer within the first 15 minutes of startup and the cluster will be broken. Unless you're updating the command-line arguments whenever the cluster is changed.

**Comment by jameshartig** [comment]
What I'm proposing in this discussion is that `UpdatePeers` is never called and once the cluster is bootstrapped the `master.peers` command line argument is ignored. In the event that you have to rescue the cluster, then you can re-bootstrap the cluster.

**Comment by kmlebedev** [comment]
>We plan to use preemptible nodes for Seaweed and will frequently have nodes being removed and added. The problem is that we don't want to programmatic have to update the -master.peers command-line

There are two update functions, one at startup, and the second when changing the leader in the first 15 minutes of the master's life.

I understood the problem and in this mode of operation was not tested.  It seems that if 15 minutes is turned into a parameter and at 0 seconds this update logic is disabled.  The remainder is to add the logic of filling the peers of the master from the raft config when the list on the command line is empty.

But when adding a new node, you need to manually add a peer via cli or implement an API for this.

**Comment by kmlebedev** [comment]
It seems that in your case it will be necessary to implement a issue
https://github.com/seaweedfs/seaweedfs/issues/3215
